src/ray/raylet/placement_group_resource_manager.cc

```diff
@@ -44,37 +44,65 @@ NewPlacementGroupResourceManager::NewPlacementGroupResourceManager(
       update_resources_(update_resources),
       delete_resources_(delete_resources) {}
 
-bool NewPlacementGroupResourceManager::PrepareBundle(
-    const BundleSpecification &bundle_spec) {
-  auto iter = pg_bundles_.find(bundle_spec.BundleId());
-  if (iter != pg_bundles_.end()) {
-    if (iter->second->state_ == CommitState::COMMITTED) {
-      // If the bundle state is already committed, it means that prepare request is just
-      // stale.
-      RAY_LOG(DEBUG) << "Duplicate prepare bundle request, skip it directly. This should "
-                        "only happen when GCS restarts.";
-      return true;
+bool NewPlacementGroupResourceManager::PrepareBundles(
+    const std::vector<std::shared_ptr<const BundleSpecification>> &bundle_specs) {
+  // Bundles that excluded bundles whose state is already committed.
+  std::vector<std::shared_ptr<const BundleSpecification>> preparing_bundles;
```

Comment:
Is there any reason why this can be just 

```
for (const auto &bundle_spec : bundle_specs) {
    PrepareBundle(bundle_spec)
}
```
?

Reply:
Got it. It makes sense. 2 especially sounds reasonable!!

Reply:
Hmm actually thinking more, I feel like this is the cleaner approach. I'd like to propose this way;

```cpp
vector prepared_bundles = []
for (const auto &bundle_spec : bundle_specs) {
    if (PrepareBundle(bundle_spec)) {
        prepared_bundles.append(bundle_spec)
    } else {
        break;
    }
}
if (len(prepared_bundles) != len(bundle_specs)) {
    for (bundle : prepared_bundles) {
        ReturnBundle(bundle);
    }
}
```

I think 1 holds for both cases. It is just the implementation details.
2 can be satisfied by this approach.

The pro of this approach is that the code is more coherent and dead obvious. It is clear we just prepare bundles and if any of them fails, we return the prepared bundles from the code. Wdyt?

Reply:
I got the approach you listed, actually, this is totally the same as what I did, but why don't we encapsulate this approach in `PgResourceManager`, I think this will be more clear.

Reply:
Maybe there was confusion? I suggested the implementation of `PrepareBundles`

Reply:
```cpp
bool NewPlacementGroupResourceManager::PrepareBundles() {
vector prepared_bundles = []
for (const auto &bundle_spec : bundle_specs) {
    if (PrepareBundle(bundle_spec)) {
        prepared_bundles.append(bundle_spec)
    } else {
        break;
    }
}
if (len(prepared_bundles) != len(bundle_specs)) {
    for (bundle : prepared_bundles) {
        ReturnBundle(bundle);
    }
}
}
```

Reply:
oh, sorry, I got it, you mean to add a new method `PrepareBundles ` which contains the original method `PrepareBundle `,  instead of modifying the original method `PrepareBundle`, right?

Reply:
yes correct!